spec/compiler/crystal/tools/unreachable_spec.cr

```diff
@@ -0,0 +1,413 @@
+      ༓def foo(&proc : ->)
+        proc.call
+      end
+
+      def bar(&proc : ->)
+        proc.call
+      end
+
+      bar {}
+      CRYSTAL
+  end
+
+  it "finds shadowed method (1)" do
+    assert_unreachable <<-CRYSTAL
+      ༓def foo
+      end
+
+      ༓def foo
+      end
+
+      ༓def bar
+      end
+
+      def bar
+      end
+
+      bar
+      CRYSTAL
+  end
+
+  it "finds method with `previous_def`" do
+    assert_unreachable <<-CRYSTAL
+      ༓def foo
+      end
+
+      ༓def foo
+        previous_def
+      end
+
+      def bar
+      end
+
+      def bar
+        previous_def
+      end
+
+      bar
+      CRYSTAL
+  end
+
+  it "finds methods called from reachable code" do
+    assert_unreachable <<-CRYSTAL
+      ༓def qux_foo
+      end
+
+      ༓def foo
+        qux_foo
+      end
+
+      def qux_bar
+      end
+
+      def bar
+        qux_bar
+      end
+
+      bar
+      CRYSTAL
+  end
+
+  it "does not finds method with `super`" do
+    assert_unreachable <<-CRYSTAL
+      class Foo
+        ༓def foo
+        end
+
+        def bar
+        end
+      end
+
+      class Qux < Foo
+        ༓def foo
+          super
+        end
+
+        def bar
+          super
+        end
+      end
+
+      Qux.new.bar
+      CRYSTAL
+  end
+
+  it "finds methods in generic type" do
+    assert_unreachable <<-CRYSTAL
+      class Foo(T)
+        ༓def foo
+          1
+        end
+
+        def bar
+          2
+        end
+      end
+
+      Foo(Int32).new.bar
+      CRYSTAL
+  end
+
+  it "finds method in abstract type" do
+    assert_unreachable <<-CRYSTAL
+      abstract class Foo
+        ༓def foo
+        end
+
+        def bar
+        end
+      end
+
+      class Baz < Foo
+      end
+
+      Baz.new.bar
+      CRYSTAL
+  end
+
+  # TODO: Should abstract Foo#bar be reported as well?
+  it "finds abstract method" do
+    assert_unreachable <<-CRYSTAL
+      abstract class Foo
+        abstract def foo
+
+        abstract def bar
+      end
+
+      class Baz < Foo
+        ༓def foo
+        end
+
+        def bar
+        end
+      end
+
+      Baz.new.bar
+      CRYSTAL
+  end
+
+  it "finds virtual method" do
+    assert_unreachable <<-CRYSTAL
+      abstract class Foo
+        ༓def foo
+        end
+
+        def bar
+        end
+      end
+
+      class Baz < Foo
+      end
+
+      class Qux < Foo
+        ༓def foo
+        end
+
+        def bar
+        end
+      end
+
+      Baz.new.as(Baz | Qux).bar
+      CRYSTAL
+  end
+
+  it "ignores autogenerated enum predicates" do
+    assert_unreachable <<-CRYSTAL
+      enum Foo
+        BAR
+        BAZ
+
+        ༓def foo
+        end
+      end
+      CRYSTAL
+  end
+
+  it "finds method called from instance variable initializer" do
+    assert_unreachable <<-CRYSTAL
+      ༓def foo
+      end
+
+      def bar
+        1
+      end
+
+      class Foo
+        @status = Bar.new
+        @other : Int32 = bar
+      end
+
+      class Bar
+        def initialize
```

Comment:
If Bar.new isn't called, would this one be reported?

Reply:
Sure. I suppose I could add another test case for `initialize`.